Convert to JSON Logging

jujaga opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
# Description

<!-- Describe your changes in detail -->
This PR changes the logging stack from morgan and npmlog to winston and express-winston to achieve pure JSON logging.
- Remove Jest maxWorkers cap
- Add JWT authorized party parsing support
- Use child component reference for logs
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
[SHOWCASE-2102](https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-2102)

## Types of changes

<!-- What types of changes does your code introduce? Uncomment all that apply: -->

<!-- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
Documentation (non-breaking change with enhancements to documentation)
Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->
Example log output:
```json
{"component":"www","message":"Listening on port 8080","level":"info","timestamp":"2021-06-10T23:59:46.832Z"}
{"component":"hello","function":"getHello","level":"info","message":"Hello World!","timestamp":"2021-06-10T23:59:48.571Z"}
{"responseTime":7,"azp":"chefs-frontend-local","contentLength":"14","httpVersion":"1.1","ip":"::ffff:127.0.0.1","method":"GET","path":"/app/api/v1/hello","statusCode":200,"userAgent":"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:89.0) Gecko/20100101 Firefox/89.0","level":"http","message":"GET /app/api/v1/hello 200 7ms","timestamp":"2021-06-10T23:59:48.581Z"}
{"component":"app","message":"Received kill signal. Shutting down...","level":"info","timestamp":"2021-06-10T23:59:57.005Z"}
```

jujaga committed:
Routine frontend npm dependency updates

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

jujaga committed:
Routine backend npm dependency updates

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

jujaga committed:
Refactor logging system to use winston from npmlog

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

jujaga committed:
Refactor http logging system to use express-winston from morgan
Fix unit test and remove test maxWorkers cap

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

jujaga committed:
Log authorized party information from JWT if it exists

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

jujaga committed:
Update stacktrace error logging
Update logs to use child component reference
Drop morganFormat config

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

loneil commented on app/src/components/log.js:
```diff
@@ -0,0 +1,59 @@
+  requestWhitelist: [], // Suppress default value output
+  responseWhitelist: [], // Suppress default value output
+  // Skip logging kube-probe requests
+  skip: (req) => req.get('user-agent') && req.get('user-agent').includes('kube-probe'),
```
👍

loneil reviewed (approved):
Tried it out locally a bit and everything looks good.

jujaga committed:
Refactor log component to use CJS style instantiation
Add NullTransport transporter definition
- This is needed to prevent winston from complaining during unit tests
Default logLevel to http
Add jsdoc and unit test coverage for log component

Signed-off-by: Jeremy Ho <jujaga@gmail.com>

MichaelChung123 reviewed (commented):
LGTM
